File: examples/codesByTopic/dotList/cylinder2d.cpp
/* This file is part of the Palabos library.
 *
 * Copyright (C) 2011-2017 FlowKit Sarl
 * Route d'Oron 2
 * 1010 Lausanne, Switzerland
 * E-mail contact: contact@flowkit.com
 *
 * The most recent release of Palabos can be downloaded at 
 * <http://www.palabos.org/>
 *
 * The library Palabos is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * The library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "palabos2D.h"
#include "palabos2D.hh"
#include "poiseuille.h"
#include "poiseuille.hh"
#include <vector>
#include <cmath>
#include <iostream>
#include <fstream>
#include <iomanip>

using namespace plb;
using namespace plb::descriptors;
using namespace std;

typedef double T;
#define DESCRIPTOR D2Q9Descriptor

/// A functional, used to initialize a pressure boundary to constant density
template<typename T>
class ConstantDensity {
public:
    ConstantDensity(T density_)
        : density(density_)
    { }
    T operator()(plint iX, plint iY) const {
        return density;
    }
private:
    T density;
};

/// A functional, used to instantiate bounce-back nodes at the locations of the cylinder
void cylinderSetup( MultiBlockLattice2D<T,DESCRIPTOR>& lattice,
                    IncomprFlowParam<T> const& parameters,
                    OnLatticeBoundaryCondition2D<T,DESCRIPTOR>& boundaryCondition )
{
    const plint nx = parameters.getNx();
    const plint ny = parameters.getNy();

    Box2D outlet(nx-1,nx-1, 1,ny-2);

    // Create Velocity boundary conditions everywhere
    boundaryCondition.setVelocityConditionOnBlockBoundaries (
            lattice, Box2D(0, nx-1, 0, 0) );
    boundaryCondition.setVelocityConditionOnBlockBoundaries (
            lattice, Box2D(0, nx-1, ny-1, ny-1) );
    boundaryCondition.setVelocityConditionOnBlockBoundaries (
            lattice, Box2D(0,0, 1,ny-2) );
    // .. except on right boundary, where we prefer a fixed-pressure condition.
    boundaryCondition.setPressureConditionOnBlockBoundaries (
            lattice, outlet );

    setBoundaryVelocity (
            lattice, lattice.getBoundingBox(),
            PoiseuilleVelocity<T>(parameters) );
    setBoundaryDensity (
            lattice, outlet,
            ConstantDensity<T>(1.) );
    initializeAtEquilibrium (
            lattice, lattice.getBoundingBox(),
            PoiseuilleVelocityAndDensity<T,DESCRIPTOR>(parameters) );

    plint cx = nx/4;
    plint cy = ny/2+2;
    plint r  = cy/4;
    DotList2D cylinderShape;
    for (plint iX=0; iX<nx; ++iX) {
        for (plint iY=0; iY<ny; ++iY) {
            if ( (iX-cx)*(iX-cx) + (iY-cy)*(iY-cy) < r*r ) {
                cylinderShape.addDot(Dot2D(iX,iY));
            }
        }
    }
    defineDynamics(lattice, cylinderShape, new BounceBack<T,DESCRIPTOR>);

    lattice.initialize();
}

void writeGifs(MultiBlockLattice2D<T,DESCRIPTOR>& lattice, plint iter)
{
    const plint imSize = 600;

    ImageWriter<T> imageWriter("leeloo");
    imageWriter.writeScaledGif(createFileName("u", iter, 6),
                               *computeVelocityNorm(lattice),
                               imSize, imSize );
}

int main(int argc, char* argv[]) {
    plbInit(&argc, &argv);

    global::directories().setOutputDir("./tmp/");

    IncomprFlowParam<T> parameters(
            (T) 1e-2,  // uMax
            (T) 100.,  // Re
            64,        // N
            5.,        // lx
            1.         // ly 
    );
    const T logT     = (T)0.01;
    const T imSave   = (T)0.02;
    const T maxT     = (T)1.1;

    writeLogFile(parameters, "Poiseuille flow");

    MultiBlockLattice2D<T, DESCRIPTOR> lattice (
            parameters.getNx(), parameters.getNy(),
            new BGKdynamics<T,DESCRIPTOR>(parameters.getOmega()) );

    OnLatticeBoundaryCondition2D<T,DESCRIPTOR>*
        boundaryCondition = createLocalBoundaryCondition2D<T,DESCRIPTOR>();

    cylinderSetup(lattice, parameters, *boundaryCondition);

    // Main loop over time iterations.
    for (plint iT=0; iT*parameters.getDeltaT()<maxT; ++iT) {
        // At this point, the state of the lattice corresponds to the
        //   discrete time iT. However, the stored averages (getStoredAverageEnergy
        //   and getStoredAverageEnergy) correspond to the previous time iT-1.

       if (iT%parameters.nStep(imSave)==0) {
            pcout << "Saving Gif ..." << endl;
            writeGifs(lattice, iT);
        }

        if (iT%parameters.nStep(logT)==0) {
            pcout << "step " << iT
                  << "; t=" << iT*parameters.getDeltaT();
        }

        // Lattice Boltzmann iteration step.
        lattice.collideAndStream();

        // At this point, the state of the lattice corresponds to the
        //   discrete time iT+1, and the stored averages are upgraded to time iT.
        if (iT%parameters.nStep(logT)==0) {
            pcout << "; av energy ="
                  << setprecision(10) << getStoredAverageEnergy<T>(lattice)
                  << "; av rho ="
                  << getStoredAverageDensity<T>(lattice) << endl;
        }
    }
    delete boundaryCondition;
}

File: jlabos/examples/Cylinder2d/Cylinder2d.java
/* This file is part of the Palabos library.
 *
 * Copyright (C) 2011-2017 FlowKit Sarl
 * Route d'Oron 2
 * 1010 Lausanne, Switzerland
 * E-mail contact: contact@flowkit.com
 *
 * The most recent release of Palabos can be downloaded at 
 * <http://www.palabos.org/>
 *
 * The library Palabos is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * The library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
 * Main author: Yann Sagon <yann.sagon@unige.ch>
 */

public class Cylinder2d{
   public static void main(String[] argv){
      jlabos.JlabosBase base = jlabos.JlabosBase.getSingletonObject();

      int lx,ly;                     // Dimensions of the system in lattice units.
      lx = 800;
      ly = 120;       
      double dx     = 1./ly;         // Discrete space step (ly is the reference length).
      int  nx,ny;                    // Lattice resolution.
      nx = lx+1;
      ny = ly+1;    
      double uMax   = 0.02;          // Maximum velocity in lattice units.
      double Re     = 200.;          // Reynolds number (wrt ly).
      double nu     = uMax*ly / Re;  // Viscosity in lattice units.
      double tau    = 3*nu+0.5;      // Relaxation time.
      long maxIter  = 4000;          // Maximum number of iterations.
      long statIter = 1000;          // Intervals at which some statistics is printed.

      jlabos.core.plbInit();

      jlabos.Box2D myDomain = new jlabos.Box2D(0, lx, 0, ly);

      // Allocate memory for the block-lattice, and assign the dynamics.
      jlabos.Box2D domain = new jlabos.Box2D(0, lx, 0, ly);
      jlabos.MultiBlockLattice2D lattice = new jlabos.MultiBlockLattice2D(domain, jlabos.Descriptor.D2Q9, new jlabos.BGK(1./tau));

      jlabos.Utils.pout(jlabos.MultiBlockUtils.info(lattice));

      // Here, x and y are (parallelized) matrices. It's similar to Matlab's meshgrid.
      jlabos.MultiNTensorField2DInt x = lattice.meshGrid(0);
      jlabos.MultiNTensorField2DInt y = lattice.meshGrid(1);

      jlabos.MultiNTensorField2DDouble xdouble = x.toDouble();
      jlabos.MultiNTensorField2DDouble ydouble = y.toDouble();

      // Compute pressure and velocity according to the analytical Poiseuille solution.
      jlabos.MultiNTensorField2DDouble pois_pressure = xdouble.subtract(new double[] {lx/2.0}).multiply(new double[] { 8*nu*uMax/(ly*ly)}); 

      jlabos.MultiNTensorField2DDouble pois_velocity = new jlabos.MultiNTensorField2DDouble(domain, 2);

      pois_velocity.setToConstant(domain, new double[] {0.,0.});

      jlabos.MultiNTensorField2DDouble pois_res = (ydouble.subtract(ydouble.multiply(ydouble).multiply(new double[] {dx}))).multiply(new double[] {4.*uMax*dx});

      pois_velocity.setComponent(0, pois_res, domain);

      // Use a regularized boundary condition. Inlet and outlet are Poiseuille profile.
      jlabos.OuterBoxBC.setVelocityCondition(lattice, domain);

      lattice.setBoundaryVelocity(pois_velocity, domain);

      lattice.initializeAtEquilibrium(pois_pressure.add(new double[] {1.}), pois_velocity, domain);

      // Define the shape of a cylinder and associate bounce-back nodes to its domain.
      int cx,cy;
      cx = nx/4;
      cy = ny/2+2;
      int r  = 10;

      jlabos.MultiNTensorField2DInt cylinder = ((xdouble.subtract(new double[] {cx})).toThePower(new double[] {2.})).
         add((ydouble.subtract(new double[] {cy}).toThePower(new double[] {2.}))).
         lessThan(new double[] {Math.pow(r,2)});


      lattice.defineDynamics(cylinder, new jlabos.BounceBack(1.));

      jlabos.Utils.pout("Iterating");

      java.util.Vector<jlabos.MultiNTensorField2DDouble> snapshots = new java.util.Vector<jlabos.MultiNTensorField2DDouble>();

      long startTime = System.currentTimeMillis()/1000;
      for (int i=0;i<maxIter;i++){
         if (i%statIter==0 && i>0){
            // Compute a velocity snapshot of the system, and keep it safe.
            snapshots.add(lattice.velocityNorm(domain));

            // Apply a reduction operation to compute the average kinetic energy.
            jlabos.Utils.pout("The energy is " + lattice.averageEnergy(domain));

            long elapsed = System.currentTimeMillis()/1000-startTime+1; //we add 1 to avoid division by zero
            double Msups = i*nx*ny / elapsed / 1e6;
            jlabos.Utils.pout("Efficiency: " + Msups + " Mega site updates per second.");
         }
         // This is the real computation: evolving through a collision-streaming cycle.
         lattice.collideAndStream();
      }

      int index = 0;
      for(jlabos.MultiNTensorField2DDouble item : snapshots){
         jlabos.Utils.saveAsJpg(item, "res"+ index);
         item.writeToFile(new String("res"+ index++ + ".txt"));
      }
   }
}
